### PR TITLE
[FIX] mail: merge duplicate help ticket sla

### PR DESCRIPTION
**Pull request by Yoann-bary**
Coupled with [77d53029986b5120500aef6dc25094588e315143](https://github.com/odoo/enterprise/pull/69352/commits/77d53029986b5120500aef6dc25094588e315143)

Steps to reproduce:
- Helpdesk > Configuration > SLA policies
- Create one if necessary or use existing
- Helpdesk > Tickets > Create
- Configure it so that it corresponds to SLA requirements
- Duplicate the ticket and merge them both (data_merge module required)

Currently helpdesk ticket uses the default merge fuction, which simply concatenates the values of merged records. This causes a few strange things to happen:

1. sla_ids which registers SLA policy types, has duplicate entries. This recordset should remain a set.

2. Followers of the merged tickets are sometimes not added to the resulting record. The default merge simply swaps the resource id with no regard for a possible duplication of unique key (res_model, res_id, partner_id), which causes the transaction to fail on repeat followers. This is a common problem with every record inheriting mail.thread which does not define its own merge function as well.

3. Keeping every SLA status record is redundant, merging tickets makes it impossbile to keep track of which task each deadline was originally related to anyway. Hence why we only keep those with the shortest deadline for each type. Priority is not a concern here since SLA policies are recomputed according to a ticket's priority rating, so we cannot have 2 SLAs with the same policy and different priorities.

opw-4037262

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/mail/models/mail_thread.py
Comment (nle-odoo) [inline addons/mail/models/mail_thread.py:4556]: can we different models that are not mail.thread (or inheriting mail.thread)? if not this doesn't seem necessary since we are in mail.thread model code that should always have "message_follower_ids" field

**Review by nle-odoo — APPROVED** [review]
seems good to me

File: addons/mail/models/mail_thread.py
Comment (Yoann-bary) [reply addons/mail/models/mail_thread.py:4556]: You're right, that check should really be in data_merge instead I'm not sure what I was thinking

Comment (Yoann-bary) [reply addons/mail/models/mail_thread.py:4556]: Or not, data merge apparently depends on mail. I'll fix the commit right away :+1:

Comment (Yoann-bary) [reply addons/mail/models/mail_thread.py]: @nle-odoo Done

File: addons/mail/models/mail_thread.py
Comment (xavierbol) [inline addons/mail/models/mail_thread.py:4559]: Seems strange to me, we have to remove all followers from the source. Is it not possible to avoid merging the followers since that method did the job? 🤔

Comment (Yoann-bary) [reply addons/mail/models/mail_thread.py:4559]: We kinda have to do this if we want to use the super later down the line. If we leave it and use _update_foreign_keys it will raise an error. This could be ignored as we don't care if the transaction fails anymore but it still flags as a failure on runbot and I'm not comfortable with the idea of sweeping the error under the rug in the code

Comment (MissingNoShiny) [reply addons/mail/models/mail_thread.py:4559]: We might never need to call super though, so maybe we could not do it while it's not necessary, and add a comment explaining that it needs to be done if a call to super is ever introduced?
I feel like it's better to avoid losing information if possible.

Comment (Yoann-bary) [reply addons/mail/models/mail_thread.py:4559]: Sorry I should clarify. The problem is a little deeper than that, we have to change the record_id of the follower in the DB but in doing so we risk duplicating the unique key for that table. The super I was referring to was _merge_method from data_merge which is always called

File: addons/mail/models/mail_thread.py
Comment (Yoann-bary) [reply addons/mail/models/mail_thread.py:4557]: :+1:

File: addons/mail/models/mail_thread.py
Comment (Yoann-bary) [reply addons/mail/models/mail_thread.py:4559]: Never mind you were right, I think this only used to happen before I removed the original merge behavior of mail.thread (Which in hindsight should have been obvious but I seem to have forgotten), we don't have to remove the followers

**Comment by Yoann-bary** [comment]
This is being moved to enterprise/data_merge

**Pull request closed without merging**
